File: sdk-typescript/scripts/version-lockstep.mjs
#!/usr/bin/env node
// One version across the five packages, siblings exact-pinned to it (S10).
//   node scripts/version-lockstep.mjs             check
//   node scripts/version-lockstep.mjs --expect V  check, and V must be the version
//   node scripts/version-lockstep.mjs --set V     rewrite every package.json (and the workspace root) to V

import { readFileSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(fileURLToPath(new URL(".", import.meta.url)), "..");
const PACKAGES = ["core", "sync", "runtime", "telemetry", "otel-bridge", "sdk"];
const args = process.argv.slice(2);
const set = args.includes("--set") ? args[args.indexOf("--set") + 1] : null;
const expect = args.includes("--expect") ? args[args.indexOf("--expect") + 1] : null;

const manifestPath = (name) => join(root, "packages", name, "package.json");
const read = (path) => JSON.parse(readFileSync(path, "utf8"));
const write = (path, value) => writeFileSync(path, JSON.stringify(value, null, 2) + "\n");

if (set) {
  if (!/^\d+\.\d+\.\d+(-[0-9A-Za-z.-]+)?$/.test(set)) throw new Error(`not a version: ${set}`);
  for (const name of PACKAGES) {
    const manifest = read(manifestPath(name));
    manifest.version = set;
    for (const dep of Object.keys(manifest.dependencies ?? {})) if (dep.startsWith("@airprompter/")) manifest.dependencies[dep] = set;
    write(manifestPath(name), manifest);
  }
  const workspace = read(join(root, "package.json"));
  workspace.version = set;
  write(join(root, "package.json"), workspace);
  // The constant the SDK reports on the heartbeat, pinned to core's package.json by packageSplit.test.ts.
  const versionTs = join(root, "packages", "core", "src", "protocol", "version.ts");
  writeFileSync(versionTs, readFileSync(versionTs, "utf8").replace(/export const SDK_VERSION = "[^"]+";/, `export const SDK_VERSION = "${set}";`));
  console.log(`version-lockstep: every package is ${set}`);
}

const problems = [];
const versions = new Map();
for (const name of PACKAGES) {
  const manifest = read(manifestPath(name));
  versions.set(name, manifest.version);
  for (const [dep, range] of Object.entries(manifest.dependencies ?? {})) {
    if (!dep.startsWith("@airprompter/")) continue;
    if (range !== manifest.version) problems.push(`${name} pins ${dep}@${range}; the lockstep version is ${manifest.version}`);
  }
}
const distinct = new Set(versions.values());
if (distinct.size !== 1) problems.push(`versions differ: ${[...versions].map(([n, v]) => `${n}=${v}`).join(", ")}`);
const version = [...distinct][0];
if (expect && version !== expect) problems.push(`the tag says ${expect}; the packages say ${version}`);
if (problems.length) {
  for (const line of problems) console.error(`version-lockstep: ${line}`);
  process.exit(1);
}
console.log(`version-lockstep: ok (${version})`);

File: sdk-python/tools/version_lockstep.py
#!/usr/bin/env python3
"""One version across the five distributions, siblings exact-pinned to it (S10).

  python tools/version_lockstep.py             check
  python tools/version_lockstep.py --expect V  check, and V must be the version
  python tools/version_lockstep.py --set V     rewrite every pyproject.toml to V
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent / "packages"
PACKAGES = ["core", "sync", "runtime", "telemetry", "agent"]
NAMES = {"core": "airprompter-agent-core", "sync": "airprompter-agent-sync", "runtime": "airprompter-agent-runtime", "telemetry": "airprompter-agent-telemetry", "agent": "airprompter-agent"}
VERSION_LINE = re.compile(r'^version = "([^"]+)"$', re.M)
PIN = re.compile(r'"(airprompter-agent(?:-[a-z]+)?)(\[[a-z,]+\])?==([^"]+)"')


def main() -> int:
    args = sys.argv[1:]
    set_to = args[args.index("--set") + 1] if "--set" in args else None
    expect = args[args.index("--expect") + 1] if "--expect" in args else None
    if set_to:
        for name in PACKAGES:
            path = ROOT / name / "pyproject.toml"
            text = VERSION_LINE.sub(f'version = "{set_to}"', path.read_text(), count=1)
            text = PIN.sub(lambda m: f'"{m.group(1)}{m.group(2) or ""}=={set_to}"', text)
            path.write_text(text)
        # The constant the SDK reports on the heartbeat, pinned to core's pyproject by test_package_split.py.
        init = ROOT / "core" / "src" / "airprompter_agent_core" / "__init__.py"
        init.write_text(re.sub(r'^SDK_VERSION = "[^"]+"$', f'SDK_VERSION = "{set_to}"', init.read_text(), count=1, flags=re.M))
        print(f"version_lockstep: every distribution is {set_to}")
    problems: list[str] = []
    versions: dict[str, str] = {}
    for name in PACKAGES:
        text = (ROOT / name / "pyproject.toml").read_text()
        m = VERSION_LINE.search(text)
        if not m:
            problems.append(f"{name}: no version")
            continue
        versions[name] = m.group(1)
        for pin in PIN.finditer(text):
            if pin.group(3) != m.group(1):
                problems.append(f"{name} pins {pin.group(1)}=={pin.group(3)}; the lockstep version is {m.group(1)}")
    distinct = set(versions.values())
    if len(distinct) != 1:
        problems.append("versions differ: " + ", ".join(f"{k}={v}" for k, v in versions.items()))
    version = next(iter(distinct)) if distinct else None
    if expect and version != expect:
        problems.append(f"the tag says {expect}; the distributions say {version}")
    if problems:
        for line in problems:
            print(f"version_lockstep: {line}", file=sys.stderr)
        return 1
    print(f"version_lockstep: ok ({version})")
    return 0


if __name__ == "__main__":
    sys.exit(main())
